### PR TITLE
Test multiple seeds

### PR DESCRIPTION
**Pull request by rikhuijzer**
The tests were failing on Julia 1.8-rc1. The cause appears to be that results from the same seeds are not always the same. In most seeds, the result are the same 5 out of 6 times or so. This wasn't tested in earlier Julia versions and now my theory is that Julia 1.8 was broken because 

1. Julia's default RNG changed between 1.7 and 1.8
2. This caused slightly different numbers to be generated
3. These numbers happened to cause the output to differ between run 1 and run 2 (whereas it used to be good for 5 out of 6 times)

I've confirmed this by adding a test over multiple seeds which makes Julia 1.7 fail. In other words, older Julia versions by chance had a RNG set which did cause a difference between run 1 and 2. By chance, the Julia 1.7 RNG did not. 

More specifically, the value for `tree.featval` is not the same between runs. I have no idea where that thing is set so I wasn't able to find the root cause yet.

### CODE REVIEW

**Comment by rikhuijzer** [comment]
Test failures are expected. Note that the tests things were already broken, but now there is a test to point it out

File: src/classification/main.jl [hunk 1]
```diff
@@ -163,16 +164,19 @@ function apply_tree(tree::LeafOrNode{S, T}, features::AbstractMatrix{S}) where {
...
+`labels` is a vector containing the distinct labels
 (eg. ["versicolor", "virginica", "setosa"]). It specifies the column ordering
-of the output matrix. """
-apply_tree_proba(leaf::Leaf{T}, features::AbstractVector{S}, labels) where {S, T} =
-    compute_probabilities(labels, leaf.values)
+of the output matrix.
+"""
+function apply_tree_proba(leaf::Leaf, features::AbstractVector, labels)
+    return compute_probabilities(labels, leaf.values)
+end
```
Comment (ablaom) [inline src/classification/main.jl:179]: This "cleanup" here and elsewhere makes sense, but I am little concerned about merge conflicts with #166 which is a rather substantial PR. If it's not too much trouble, please move these to a separate PR.

**Comment by ablaom** [comment]
Yes, I have also been noticing these inconsistent runs. I would think the solution is to replace all RNGs in testing with StableRNGs, adding StableRNGs to the test dependencies.

**Comment by ablaom** [comment]
Okay I think @yufongpeng has isolated the bug, corrected in this PR: https://github.com/yufongpeng)https://github.com/JuliaAI/DecisionTree.jl/pull/167. In `nfoldsCV` the `rng` is not being passed to calls to to train AdaStumpClassifier (but is being passed to the other classifiers). 

However, that PR is breaking, so I think we should release a patch to fix the bug before #167, and should probably include StableRNGs at the same time. @rikhuijzer Would you be willing and able to do that this week?

**Comment by rikhuijzer** [comment]
> Would you be willing and able to do that this week?

Done.

> should probably include StableRNGs at the same time

Can we do that in another PR? At this moment, there is no need for `StableRNGs`. The tests pass at all Julia versions including 1.8-rc1 (which I am running locally). 

And thanks @yufongpeng for spotting the source of the problem!

**Comment by rikhuijzer** [comment]
In 2f70ff5, the tests pass. It would indeed be better to add `StableRNGs` as you mentioned @ablaom, but I am running into new problems there. For some reason
```julia
DecisionTree.nfoldCV_forest(labels, features, 3, 2, 10; rng=StableRNG(10))
```
produces different results.

Unfortunately, I should now start working on some other stuff so let's leave that issue for another PR

File: src/classification/main.jl [hunk 2]
```diff
@@ -221,7 +222,9 @@ function build_forest(
 
     if rng isa Random.AbstractRNG
         Threads.@threads for i in 1:n_trees
-            inds = rand(rng, 1:t_samples, n_samples)
+            # The Mersenne Twister (Julia's default) is not thread-safe.
+            _rng = copy(rng)
+            inds = rand(_rng, 1:t_samples, n_samples)
```
Comment (rikhuijzer) [inline src/classification/main.jl:227]: This fixes another bug. This path wasn't tested. The path below on the integers does work

**Comment by rikhuijzer** [comment]
> For some reason
> 
> ```julia
> DecisionTree.nfoldCV_forest(labels, features, 3, 2, 10; rng=StableRNG(10))
> ```
> 
> produces different results while `rng=10` doesn't.

This was caused by incorrect usage of the RNG in combination with `@threads`. I fixed it now in `src/correlations/main.jl` and `src/regression/main.jl`. I did not fix the code duplication because I fear that it will add even more merge conflicts in the other open PRs.

**Comment by rikhuijzer** [comment]
The tests now fail because many comparisons depend on the seed?

https://memeshappen.com/meme/56256/yeah-if-we-could-get-i-t-to-fix-one-thing-without-breaking-3-other-things-that-would-be-great

**Comment by rikhuijzer** [comment]
Things are still not stable. The tests such as the ones in `regression/random` pass for me locally but fail in CI.

**Comment by ablaom** [comment]
@rikhuijzer Thanks for extra work and good catch with the further bug discovery. For now I'm going to hope you get a chance to finish this off, as it seems like a work-in-progress.

Side note: Strange, I didn't realize forest training was multithreaded. I thought `Threads` only became available in 1.3 and yet 1.0 tests are passing, which I don't understand.

**Comment by rikhuijzer** [comment]
> Things are still not stable. The tests such as the ones in `regression/random` pass for me locally but fail in CI.

Fixed in ed4602b. I forgot to also pass `StableRNG` to the data generation steps.

**Comment by rikhuijzer** [comment]
In 5a10555, I've put the old numbers back. I relaxed the bounds of the tests because sometimes the unstable tests had bad luck and the test would be outside of the bounds. That is fixed now by the stable rngs.

@ablaom I think this PR is good to go. If you approve, I would prefer a squash merge on this PR for the reasons mentioned in https://discourse.julialang.org/t/downsides-of-using-squash-merging/68317. You can decide :+1:

**Review by ablaom — APPROVED** [review]
A labour of love. Great PR, thanks.

**Comment by rikhuijzer** [comment]
> A labour of love. Great PR, thanks.

👍👍👍 Thanks
